Fix tests that fail in Django 4.2 due to accessing reverse foreign key relationship before save

gherceg opened the pull request:
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->

## Technical Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
From Django 4.1 [release notes](https://docs.djangoproject.com/en/5.0/releases/4.1/#reverse-foreign-key-changes-for-unsaved-model-instances):
> In order to unify the behavior with many-to-many relations for unsaved model instances, a reverse foreign key now raises ValueError when calling [related managers](https://docs.djangoproject.com/en/5.0/ref/models/relations/#django.db.models.fields.related.RelatedManager) for unsaved objects.

This seems to primarily be an issue in tests since it is common practice to create an object but not save it to the db in the test, potentially accessing a reverse fk relationship in that time.

I do still have a todo item which is to find all possible reverse fk relationships and see if where we attempt to access those to confirm the upgrade to 4.2 is safe.

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
N/A

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

gherceg committed:
Catch ValueError when attempting to access reverse fk relationship

gherceg committed:
Create RepeatRecord before attempting to requeue

Otherwise, a ValueError is raised because of attempting to access
reverse fk relationship before save

millerdev commented on corehq/apps/users/models_role.py:
```diff
@@ -196,7 +196,12 @@ def _clear_query_cache():
-        return [rp.as_permission_info() for rp in self.rolepermission_set.all()]
+        try:
+            return [rp.as_permission_info() for rp in self.rolepermission_set.all()]
+        except ValueError:
```
`ValueError` is a generic exception that could be raised for other reasons than accessing a reverse foreign key relationship. Is there a more specific exception that can be caught? Alternatively, would it work to wrap only `self.rolepermission_set.all()` or a subset of it in the try/except?

It is not clear to me what is actually raising the error. Is it raised by `self.rolepermission_set`, `self.rolepermission_set.all()` or on iterating over the return value of the latter? Whatever it is, if we must catch `ValueError` we should limit the try/except to the narrowest scope possible to avoid catching an error raised somewhere else.

gherceg replied:
Yeah good feedback. I'll wrap only what is needed, but unfortunately there isn't a more specific error to catch.

gherceg replied:
Sorry I force pushed because of a sloppy commit, but only changed the most recent commit that you hadn't yet reviewed: 4b84b5337ececa5f7fd83fb9f7860ce7961dbe28 is the best I could do

gherceg committed:
Only wrap what is necessary in try/except to raise ValueError

millerdev reviewed (commented):
Am I understanding correctly that this PR implements the old pre-Django-4.x behavior in places where our code accesses reverse foreign key relationships on unsaved objects?

If that's correct, I wonder if we are doing something wrong? Or maybe not wrong, but is there a better way we could be doing things? Like should we be writing our code in a way that a reverse fk would never be accessed on an unsaved object?

gherceg commented:
> Am I understanding correctly that this PR implements the old pre-Django-4.x behavior in places where our code accesses reverse foreign key relationships on unsaved objects?

Yeah my understanding is the previous behavior returns an empty queryset when accessing a reverse foreign key relationship on an unsaved object.

My overly optimistic hunch is that this only presents a problem for us in tests, because it is likely that we could create an unsaved db object (to keep SimpleTestCase) and attempt to use it in a context where only a saved object is used in production code. Obviously, I have not verified this so maybe all of that is moot. The justification for the breaking change without notice seems to be that this was never documented behavior and therefore does not need to go through the typical deprecation/removal cycle. Here's [the ticket](https://code.djangoproject.com/ticket/19580) if you want more info, but notably the last couple of comments are disgruntled developers who are in a similar situation as us.

All that to say, rather than the try/except blocks I've added here, the alternate solution would be to save objects in tests (like d3e9ba7d7ae0a659b078986541a323a88590f7f8), which entails switching some SimpleTestCase classes to TestCase.

gherceg committed:
Do not change production code to accommodate tests

Rather than catch these ValueErrors, update the tests to either avoid
needing to access the reverse foreign key relationship, or fake it

gherceg committed:
Run isort

gherceg commented:
Daniel and I offlined about this and decided that not modifying production code to accommodate this new error being raised seems like the right place to start. So errors will be loud if they do occur when this is running against django 4.2 on staging.

mjriley commented on corehq/apps/linked_domain/tests/test_updates.py:
```diff
@@ -1,27 +1,38 @@
-            default_landing_page='testpage',
-            is_non_admin_editable=False,
-        )
+        source_role_json = RoleCopyTests._get_source_role_as_json()
```
I don't think this is blocking, but this does change the way the test is read. I prefer tests that contain all information necessary to understand them within their bodies. The new helper function hides that information within its body. Given that we're verifying the `name`, `default_landing_page`, and `is_non_admin_editable` properties, this would seem like a more readable test:
```suggestion
        source_role_json = RoleCopyTests._get_source_role_as_json(
		  name='test',
		  default_landing_page='testpage',
		  is_non_admin_editable=False
		)
```

gherceg replied:
Yeah that is fair. I initially didn't even have the helper, and really only added it with the thought of making it as easy as possible for those in the future who may need to update these tests. It also does hide the fields that the test doesn't care about which is nice, so I like your solution which feels like a good middle ground.

gherceg replied:
7a3f0f43913b681e887329c04a0ffc30309ab39b

mjriley commented on corehq/apps/linked_domain/tests/test_updates.py:
```diff
@@ -1,27 +1,38 @@
+            "default_landing_page": "testpage",
+            "is_non_admin_editable": False,
+        }
+        # add/override specified kwargs
+        for key, val in kwargs.items():
+            source_role_json[key] = val
```
Any reason for the loop over:
```suggestion
         source_role_json.update(kwargs)
```

gherceg replied:
Nope! I'll make that change as well.

gherceg replied:
7a3f0f43913b681e887329c04a0ffc30309ab39b

mjriley reviewed (approved):
Appreciate the changes. Only non-blocking feedback

gherceg committed:
Address feedback

- Explicitly define fields that are relevant to the test when creating
  json
- no need to iterate over kwargs, just use .update(...)

mjriley reviewed (approved):
Thanks for the changes
